### PR TITLE
Brainpool curves

### PR DESCRIPTION
**Pull request by reaperhulk**
Add the brainpool curves required for German smart meters and no more than that.

supersedes #4021

### CODE REVIEW

**Comment by jlaine** [comment]
Speaking of docs, I thought we were now using docstrings for classes?

**Comment by reaperhulk** [comment]
You might be thinking of pyopenssl?

**Comment by jlaine** [comment]
Er. you're right, my bad

File: tests/hazmat/primitives/test_ec.py [hunk 1]
```diff
@@ -1101,6 +1101,31 @@ def test_key_exchange_with_vectors(self, backend, vector):
...
+                int(vector['x_qa'], 16), int(vector['y_qa'], 16), curve()
+            )
+        ).private_key(backend)
+        peer = ec.EllipticCurvePrivateNumbers(
+            int(vector['db'], 16),
+            ec.EllipticCurvePublicNumbers(
+                int(vector['x_qb'], 16), int(vector['y_qb'], 16), curve()
+            )
+        ).private_key(backend)
+        shared_secret = key.exchange(ec.ECDH(), peer.public_key())
```
Comment (alex) [inline tests/hazmat/primitives/test_ec.py]: Should we test both directions of doing the kex?

File: docs/hazmat/primitives/asymmetric/ec.rst [hunk 2]
```diff
@@ -414,6 +414,28 @@ All named curves are instances of :class:`EllipticCurve`.
     SECG curve ``secp256k1``.
 
 
+.. class:: BrainpoolP256R1
+
+    .. versionadded:: 2.2
+
+    Brainpool curve specified in :rfc:`5639`. These curves are not commonly
+    used but are available for interoperating with existing systems.
```
Comment (alex) [inline docs/hazmat/primitives/asymmetric/ec.rst]: Let's go stronger: These curves are discouraged for new systems, in favor of X25519, but...

**Pull request merged**